src/Microsoft.VisualStudio.Editor.Razor/DefaultEditorSettingsManager.cs

```diff
@@ -12,26 +13,95 @@ namespace Microsoft.VisualStudio.Editor.Razor
+        private EditorSettings _settings;
 
         [ImportingConstructor]
-        public DefaultEditorSettingsManager(VisualStudioWorkspaceAccessor workspaceAccessor)
+        public DefaultEditorSettingsManager(ForegroundDispatcher foregroundDispatcher)
         {
-            var razorLanguageServices = workspaceAccessor.Workspace.Services.GetLanguageServices(RazorLanguage.Name);
-            _editorSettingsManager = razorLanguageServices.GetRequiredService<EditorSettingsManagerInternal>();
+            _foregroundDispatcher = foregroundDispatcher;
+            _settings = EditorSettings.Default;
+            _changedHandlers = new List<WeakReference<EventHandler<EditorSettingsChangedEventArgs>>>();
         }
 
         public override event EventHandler<EditorSettingsChangedEventArgs> Changed
         {
-            add => _editorSettingsManager.Changed += value;
-            remove => _editorSettingsManager.Changed -= value;
+            add
+            {
+                _foregroundDispatcher.AssertForegroundThread();
+
+                _changedHandlers.Add(new WeakReference<EventHandler<EditorSettingsChangedEventArgs>>(value));
+            }
+            remove
+            {
+                _foregroundDispatcher.AssertForegroundThread();
+
+                for (var i = 0; i < _changedHandlers.Count; i++)
+                {
+                    if (_changedHandlers[i].TryGetTarget(out var target) && target == value)
+                    {
+                        _changedHandlers.RemoveAt(i);
+                        return;
+                    }
+                }
+            }
         }
 
-        public override EditorSettings Current => _editorSettingsManager.Current;
+        public override EditorSettings Current
+        {
+            get
+            {
+                lock (SettingsAccessorLock)
+                {
+                    return _settings;
+                }
+            }
+        }
 
-        public override void Update(EditorSettings updateSettings)
+        public override void Update(EditorSettings updatedSettings)
         {
-            _editorSettingsManager.Update(updateSettings);
+            if (updatedSettings == null)
+            {
+                throw new ArgumentNullException(nameof(updatedSettings));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            lock (SettingsAccessorLock)
+            {
+                if (!_settings.Equals(updatedSettings))
+                {
+                    _settings = updatedSettings;
+                    OnChanged();
+                }
+            }
+        }
+
+        private void OnChanged()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var handlersToInvoke = new List<EventHandler<EditorSettingsChangedEventArgs>>();
+            for (var i = 0; i < _changedHandlers.Count; i++)
+            {
+                if (_changedHandlers[i].TryGetTarget(out var handler))
```

Comment:
I'm not super stoked on the idea, but we could patternize this into a reusable weakevent class. It's a fairly common pattern thanks to WPF, but we don't want to depend on that.